### PR TITLE
Convert tests & examples to unit tests

### LINKED ISSUE #65: Port integration tests to unit-tests
from `src/test/integration/`  

- [ ] PyRegion
- [ ] CppRegion
- [ ] ConnectionPerformance-Test
- [ ] move other `src/examples/` to unit tests

this will
- simplify cmake
- better coverage
- easier to run tests in CI

### PR DESCRIPTION
**Pull request by breznak**
moves a region demo from examples to unit tests. 

- [x] fix file location
- [x] examples/region/HelloRegion -> HelloRegionTest
- [x] examples/algorithm/HelloSP_TP
- [x] tests/integration/PyRegion
- [x] tests/integration/CppRegion
- [x] tests/integration/ConnectionsPerformanceTest
- [ ] moving the Py tests in nupic.cpp/bindings/py so all tests were together

Fixes #65

### CODE REVIEW

File: src/test/unit/engine/HelloRegionTest.cpp [hunk 1]
```diff
@@ -50,7 +55,7 @@ int main(int argc, const char *argv[]) {
 
   // Load data
   std::string path =
-      Path::makeAbsolute("../../../src/examples/regions/Data.csv");
+      Path::makeAbsolute("../release/data/Data.csv"); //FIXME use path relative to CMake's nupic root
```
Comment (breznak) [inline src/test/unit/engine/HelloRegionTest.cpp]: FIXME: how do I address this file somehow absolutely? If this is run by hand from ./build/scripts/, this works; if ran from make tests_all, it has apparently a different relative location. Do we have some variable that addresses the root of the repo?

File: circle.yml [hunk 2]
```diff
@@ -97,11 +97,7 @@ jobs:
           command: |
             mkdir -p tests/cpp
             pushd build/release/bin
-            ./cpp_region_test
             ./py_region_test
-            ./helloregion
-            ./hello_sp_tp
-            ./prototest
```
Comment (breznak) [inline circle.yml]: cleaned up CI stuff 
(py_region is TODO)

File: src/CMakeLists.txt [hunk 3]
```diff
@@ -457,57 +443,6 @@ add_custom_target(tests_py_region
...
-                      ${src_common_test_exe_libs})
-set_target_properties(${src_executable_connectionsperformancetest}
-                      PROPERTIES COMPILE_FLAGS ${src_compile_flags})
-set_target_properties(${src_executable_connectionsperformancetest}
-                      PROPERTIES LINK_FLAGS "${INTERNAL_LINKER_FLAGS_OPTIMIZED}")
-add_custom_target(tests_connections_performance
-                  COMMAND ${src_executable_connectionsperformancetest}
-                  DEPENDS ${src_executable_connectionsperformancetest}
-                  COMMENT "Executing test ${src_executable_connectionsperformancetest}"
-                  VERBATIM)
```
Comment (breznak) [inline src/CMakeLists.txt]: cleaned up CmakeLists

File: src/CMakeLists.txt [hunk 4]
```diff
@@ -518,14 +453,19 @@ add_executable(${src_executable_gtests}
                test/unit/algorithms/Cells4Test.cpp
                test/unit/algorithms/CondProbTableTest.cpp
                test/unit/algorithms/ConnectionsTest.cpp
+	       test/unit/algorithms/ConnectionsPerformanceTest.cpp
+	       test/unit/algorithms/HelloSPTPTest.cpp
```
Comment (breznak) [inline src/CMakeLists.txt]: add all examples/ and tests/integration/ as proper unit-tests

File: src/test/unit/algorithms/ConnectionsPerformanceTest.cpp [hunk 5]
```diff
@@ -280,9 +251,46 @@ vector<CellIdx> ConnectionsPerformanceTest::computeSPWinnerCells(
...
+  srand(SEED);
+  t = testing::ConnectionsPerformanceTest();
+}
+
+/**
+ * Tests typical usage of Connections with Temporal Memory.
+ */
+TEST(ConnectionsPerformanceTest, testTM) {
+	auto tim = t.runTemporalMemoryTest(2048, 40, 20, 100, "temporal memory");
+	ASSERT_LE(tim, 1.5f); //there are times, we must be better. Bit underestimated for slow CI
```
Comment (breznak) [inline src/test/unit/algorithms/ConnectionsPerformanceTest.cpp]: Idea is to keep an eye on the performance/speed by the CI. If someone severely breaks the performance, these tests should trigger! 

Problem is the CI is slow compared to normal PC, and that the load there varies -> we'll have to have a very large "buffer" for the slowest CI. 

If someone improves performance significantly, these numbers should be updated (lowered) to reflect that. 
Once set, we ideally never should increase the numbers (means losing performance).

File: src/test/unit/algorithms/Serialization.cpp [hunk 6]
```diff
@@ -34,11 +36,13 @@
...
 #include <nupic/utils/Random.hpp>
 
+namespace testing {
+
 using namespace std;
 using namespace nupic;
 using namespace nupic::algorithms::spatial_pooler;
 
-void testSP() {
+TEST(Serialization, testSP) {
```
Comment (breznak) [inline src/test/unit/algorithms/Serialization.cpp]: I think youve removed/changed this file (former prototest), whichever comes first, we'll have to update the merge conflict.

File: src/test/unit/engine/CppRegionTest.cpp [hunk 7]
```diff
@@ -43,30 +45,24 @@
 #include <string>
 #include <vector>
 
-bool ignore_negative_tests = false;
-#define SHOULDFAIL(statement)                                                  \
```
Comment (breznak) [inline src/test/unit/engine/CppRegionTest.cpp]: remove this, other for removel too, in other PR.

File: src/test/unit/engine/CppRegionTest.cpp [hunk 8]
```diff
@@ -43,30 +45,24 @@
...
 
 using namespace nupic;
+using std::exception;
 
 bool verbose = false;
 
+
 struct MemoryMonitor {
-  MemoryMonitor() { OS::getProcessMemoryUsage(initial_vmem, initial_rmem); }
+  MemoryMonitor(const size_t count) : maxIter(count) {
```
Comment (breznak) [inline src/test/unit/engine/CppRegionTest.cpp]: this was never used in CI, I've fixed it

File: src/test/unit/engine/CppRegionTest.cpp [hunk 9]
```diff
@@ -110,15 +133,18 @@ void testCppInputOutputAccess(Region *level1) {
...
+
+
+TEST(CppRegionTest, testCppLinkingFanIn) {
   Network net = Network();
 
   Region *region1 = net.addRegion("region1", "TestNode", "");
   Region *region2 = net.addRegion("region2", "TestNode", "");
-  net.link("region1", "region2", linkPolicy, linkParams);
+
+  net.link("region1", "region2", "TestFanIn2", ""); //the only change testCppLinking* is here
```
Comment (breznak) [inline src/test/unit/engine/CppRegionTest.cpp]: this is ugly, I had to duplicate the 2 methods, and only this line (link) changes. Didn't know other way how to do it for gtest.

File: src/test/unit/engine/CppRegionTest.cpp [hunk 10]
```diff
@@ -143,28 +169,110 @@ void testCppLinking(std::string linkPolicy, std::string linkParams) {
...
+
+
+
+TEST(CppRegionTest, testCppLinkingUniformLink) {
+  Network net = Network();
+
+  Region *region1 = net.addRegion("region1", "TestNode", "");
+  Region *region2 = net.addRegion("region2", "TestNode", "");
+
+  net.link("region1", "region2", "UniformLink", "{mapping: in, rfSize: [2]}"); //the only change testCppLinking* is here
```
Comment (breznak) [inline src/test/unit/engine/CppRegionTest.cpp]: the duplicated method

File: src/test/unit/engine/CppRegionTest.cpp [hunk 11]
```diff
@@ -311,125 +422,81 @@ int realmain(bool leakTest) {
...
+//  EXPECT_NO_THROW(testCppLinking("UniformLink", "{mapping: in, rfSize: [2]}"));
+//  EXPECT_NO_THROW(testYAML());
 
-  return 0;
 }
 
-int main(int argc, char *argv[]) {
 
+
+TEST(DISABLED_CppRegionTest, memLeak) { //FIXME this mem leak test is newly fixed, but catches error -> need to fix code
```
Comment (breznak) [inline src/test/unit/engine/CppRegionTest.cpp]: the new test for memory leak, had to disable because it currrently fails (and not aim of this PR)
TLDR: we have memory leaks,

File: src/test/unit/engine/HelloRegionTest.cpp [hunk 12]
```diff
@@ -20,19 +20,23 @@
...
 
 #include <nupic/engine/Network.hpp>
 #include <nupic/engine/Region.hpp>
+
 #include <nupic/ntypes/ArrayRef.hpp>
 #include <nupic/ntypes/Dimensions.hpp>
 #include <nupic/os/Path.hpp>
-#include <nupic/utils/Log.hpp>
+
+namespace testing {
```
Comment (breznak) [inline src/test/unit/engine/HelloRegionTest.cpp]: let's try to put all new tests in "testing" now,ok?

**Comment by breznak** [comment]
This is the error on Windows, have you seen it? 
https://ci.appveyor.com/project/breznak/nupic-cpp/build/0.3.0.172#L256

**Comment by breznak** [comment]
- testTP fails in Travis : https://travis-ci.org/htm-community/nupic.cpp/builds/422452622#L6784 
  120s instead of 80 :P

**Comment by breznak** [comment]
@dkeeney what was the idea about 
> moving the Py tests in nupic.cpp/bindings/py so all tests were together 

I cannot find it now and py tests seem to be indeed in bindings/py/ path

**Comment by breznak** [comment]
Stuck on one failing test on OSX: 

```
[----------] 1 test from HelloRegionTest
[ RUN      ] HelloRegionTest.demo
Setting region dimensions[1]
Loading data from /Users/distiller/numenta/nupic.core/build/release/bin/../data/Data.csv
In VectorFileSensor /Users/distiller/numenta/nupic.core/build/release/bin/../data/Data.csv
INFO:   appendFile - need to reset scale and offset vectors.
Read 4 vectors
Initializing network
Compute
  0    -2.31584e+77
/Users/distiller/numenta/nupic.core/src/test/unit/engine/HelloRegionTest.cpp:86: Failure
Value of: 0
Expected: buffer[0]
Which is: -inf
In VectorFileSensor /Users/distiller/numenta/nupic.core/build/release/bin/../data/Data.csv
INFO:   appendFile - need to reset scale and offset vectors.
Read 4 vectors
 buffer -2.31584e+77 buffer2: -2.31584e+77
[  FAILED  ] HelloRegionTest.demo (1 ms)
[----------] 1 test from HelloRegionTest (1 ms total)
```
Any idea?

**Comment by breznak** [comment]
@dkeeney please review and merge, if all tests pass

**Comment by breznak** [comment]
> I am a little concerned that the memory leak detection might not actually be working. 

That's quite possible. I just made to original, dusty code to compile. It has never been used systematically in the past. #76 should look at it in detail. Although I'd agree to possibly leave it to external tools - valgrind, etc.
